File: python/relayer-graph/src/relayer_graph/query.py
"""Typed public wire contract for bounded graph search."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, Mapping, TypedDict, Union


class GraphQueryBudget(TypedDict, total=False):
    queryBytes: int
    astDepth: int
    variables: int
    patternParts: int
    traversalHops: int
    examinedExpansions: int
    intermediateRows: int
    wallTimeMs: int
    resultRows: int
    encodedResultBytes: int


class GraphQueryScalarTypeDescriptor(TypedDict):
    kind: Literal["null", "boolean", "integer", "float", "string", "node", "layer", "relationship", "path"]


class GraphQueryListTypeDescriptor(TypedDict):
    kind: Literal["list"]
    elementType: "GraphQueryTypeDescriptor"


class GraphQueryRecordTypeField(TypedDict):
    name: str
    type: "GraphQueryTypeDescriptor"


class GraphQueryRecordTypeDescriptor(TypedDict):
    kind: Literal["record"]
    fields: list[GraphQueryRecordTypeField]


GraphQueryTypeDescriptor = Union[GraphQueryScalarTypeDescriptor,
                                 GraphQueryListTypeDescriptor,
                                 GraphQueryRecordTypeDescriptor]


class GraphQueryNullValue(TypedDict):
    type: Literal["null"]


class GraphQueryBooleanValue(TypedDict):
    type: Literal["boolean"]
    value: bool


class GraphQueryIntegerValue(TypedDict):
    type: Literal["integer"]
    # Signed 64-bit integers intentionally remain decimal strings in Python.
    value: str


class GraphQueryFloatValue(TypedDict):
    type: Literal["float"]
    value: float


class GraphQueryStringValue(TypedDict):
    type: Literal["string"]
    value: str


class GraphQueryProperty(TypedDict):
    name: str
    value: "GraphQueryValue"


class GraphQueryNodeValue(TypedDict):
    type: Literal["node"]
    id: str
    kind: Literal["Content"]
    properties: list[GraphQueryProperty]


class GraphQueryLayerValue(TypedDict):
    type: Literal["layer"]
    id: str
    kind: Literal["Layer"]
    properties: list[GraphQueryProperty]


class GraphQueryRelationshipValue(TypedDict):
    type: Literal["relationship"]
    id: str
    kind: Literal["CONNECTED", "CONTAINS", "EXPANDS", "REFERENCES"]
    start: str
    end: str
    directed: bool
    properties: list[GraphQueryProperty]


class GraphQueryPathValue(TypedDict):
    type: Literal["path"]
    vertices: list[Union[GraphQueryNodeValue, GraphQueryLayerValue]]
    relationships: list[GraphQueryRelationshipValue]


class GraphQueryListValue(TypedDict):
    type: Literal["list"]
    elementType: GraphQueryTypeDescriptor
    values: list["GraphQueryValue"]


class GraphQueryRecordField(TypedDict):
    name: str
    value: "GraphQueryValue"


class GraphQueryRecordValue(TypedDict):
    type: Literal["record"]
    fields: list[GraphQueryRecordField]


GraphQueryValue = Union[GraphQueryNullValue, GraphQueryBooleanValue,
                        GraphQueryIntegerValue, GraphQueryFloatValue,
                        GraphQueryStringValue, GraphQueryNodeValue,
                        GraphQueryLayerValue, GraphQueryRelationshipValue,
                        GraphQueryPathValue, GraphQueryListValue,
                        GraphQueryRecordValue]


class GraphSearchResult(TypedDict):
    queryContractVersion: Literal[1]
    columns: list[str]
    rows: list[list[GraphQueryValue]]
    truncated: bool


class GraphSearchTarget(TypedDict):
    """A known logical dataset selector. It conveys no authority."""

    scope: Literal["thread", "project"]
    id: int


@dataclass(frozen=True, slots=True)
class GraphSearchRequest:
    """Search input; omission selects the current interaction's thread."""

    query: str
    parameters: Mapping[str, GraphQueryValue] = field(default_factory=dict)
    budget: GraphQueryBudget = field(default_factory=GraphQueryBudget)
    target: GraphSearchTarget | None = None
    query_contract_version: Literal[1] = 1

    def to_wire(self) -> dict[str, object]:
        request: dict[str, object] = {
            "queryContractVersion": self.query_contract_version,
            "query": self.query,
            "parameters": dict(self.parameters),
            "budget": dict(self.budget),
        }
        if self.target is not None:
            request = {
                "queryContractVersion": self.query_contract_version,
                "target": {"scope": self.target["scope"], "id": self.target["id"]},
                "query": self.query,
                "parameters": dict(self.parameters),
                "budget": dict(self.budget),
            }
        return request

File: packages/graph-client/src/query.ts
export type GraphQueryPhase =
  | "envelope"
  | "parse"
  | "plan"
  | "authorize"
  | "execute"
  | "normalize"
  | "encode";

export type GraphQueryCode =
  | "invalid_request"
  | "unsupported_query_contract_version"
  | "query_bytes_exceeded"
  | "query_syntax_invalid"
  | "query_construct_forbidden"
  | "query_construct_unsupported"
  | "unknown_label"
  | "unknown_relationship_type"
  | "unknown_property"
  | "dynamic_schema_forbidden"
  | "query_type_mismatch"
  | "duplicate_output_column"
  | "invalid_aggregate"
  | "ast_depth_exceeded"
  | "variable_limit_exceeded"
  | "pattern_part_limit_exceeded"
  | "traversal_limit_exceeded"
  | "row_limit_exceeded"
  | "inaccessible_or_missing"
  | "scope_not_granted"
  | "foreign_draft"
  | "query_cancelled"
  | "wall_time_exceeded"
  | "examined_expansions_exceeded"
  | "intermediate_rows_exceeded"
  | "invalid_engine_value"
  | "heterogeneous_list"
  | "integer_overflow"
  | "duplicate_record_field"
  | "result_row_too_large";

export interface GraphQueryBudget {
  readonly queryBytes?: number;
  readonly astDepth?: number;
  readonly variables?: number;
  readonly patternParts?: number;
  readonly traversalHops?: number;
  readonly examinedExpansions?: number;
  readonly intermediateRows?: number;
  readonly wallTimeMs?: number;
  readonly resultRows?: number;
  readonly encodedResultBytes?: number;
}

export type GraphQueryTypeDescriptor =
  | { readonly kind: "null" | "boolean" | "integer" | "float" | "string" | "node" | "layer" | "relationship" | "path" }
  | { readonly kind: "list"; readonly elementType: GraphQueryTypeDescriptor }
  | { readonly kind: "record"; readonly fields: readonly GraphQueryRecordTypeField[] };

export interface GraphQueryRecordTypeField {
  readonly name: string;
  readonly type: GraphQueryTypeDescriptor;
}

export interface GraphQueryNullValue { readonly type: "null" }
export interface GraphQueryBooleanValue { readonly type: "boolean"; readonly value: boolean }
/** Signed 64-bit integers intentionally remain decimal strings across JSON runtimes. */
export interface GraphQueryIntegerValue { readonly type: "integer"; readonly value: string }
export interface GraphQueryFloatValue { readonly type: "float"; readonly value: number }
export interface GraphQueryStringValue { readonly type: "string"; readonly value: string }

export interface GraphQueryProperty {
  readonly name: string;
  readonly value: GraphQueryValue;
}

export interface GraphQueryNodeValue {
  readonly type: "node";
  readonly id: string;
  readonly kind: "Content";
  readonly properties: readonly GraphQueryProperty[];
}

export interface GraphQueryLayerValue {
  readonly type: "layer";
  readonly id: string;
  readonly kind: "Layer";
  readonly properties: readonly GraphQueryProperty[];
}

export interface GraphQueryRelationshipValue {
  readonly type: "relationship";
  readonly id: string;
  readonly kind: "CONNECTED" | "CONTAINS" | "EXPANDS" | "REFERENCES";
  readonly start: string;
  readonly end: string;
  readonly directed: boolean;
  readonly properties: readonly GraphQueryProperty[];
}

export interface GraphQueryPathValue {
  readonly type: "path";
  readonly vertices: readonly (GraphQueryNodeValue | GraphQueryLayerValue)[];
  readonly relationships: readonly GraphQueryRelationshipValue[];
}

export interface GraphQueryListValue {
  readonly type: "list";
  readonly elementType: GraphQueryTypeDescriptor;
  readonly values: readonly GraphQueryValue[];
}

export interface GraphQueryRecordField {
  readonly name: string;
  readonly value: GraphQueryValue;
}

export interface GraphQueryRecordValue {
  readonly type: "record";
  readonly fields: readonly GraphQueryRecordField[];
}

export type GraphQueryValue =
  | GraphQueryNullValue
  | GraphQueryBooleanValue
  | GraphQueryIntegerValue
  | GraphQueryFloatValue
  | GraphQueryStringValue
  | GraphQueryNodeValue
  | GraphQueryLayerValue
  | GraphQueryRelationshipValue
  | GraphQueryPathValue
  | GraphQueryListValue
  | GraphQueryRecordValue;

/**
 * Public current-thread search envelope. It deliberately has no target,
 * project, thread, scope, permit, credential, or other authority selector.
 */
export interface GraphSearchRequest {
  readonly queryContractVersion: 1;
  readonly query: string;
  readonly parameters: Readonly<Record<string, GraphQueryValue>>;
  readonly budget: GraphQueryBudget;
}

export interface GraphSearchResult {
  readonly queryContractVersion: 1;
  readonly columns: readonly string[];
  readonly rows: readonly (readonly GraphQueryValue[])[];
  readonly truncated: boolean;
}

export interface GraphSearchOptions {
  readonly signal?: AbortSignal;
}

export interface GraphQueryErrorBody {
  readonly error?: {
    readonly code?: string;
    readonly phase?: string;
    readonly path?: string;
    readonly message?: string;
  };
}

export class GraphQueryError extends Error {
  constructor(
    readonly status: number,
    readonly code: GraphQueryCode,
    readonly phase: GraphQueryPhase,
    readonly path: string,
    message: string,
  ) {
    super(message);
    this.name = "GraphQueryError";
  }
}

const QUERY_CODE_PHASES: Readonly<Record<GraphQueryCode, GraphQueryPhase>> = {
  invalid_request: "envelope",
  unsupported_query_contract_version: "envelope",
  query_bytes_exceeded: "envelope",
  query_syntax_invalid: "parse",
  query_construct_forbidden: "parse",
  query_construct_unsupported: "parse",
  unknown_label: "plan",
  unknown_relationship_type: "plan",
  unknown_property: "plan",
  dynamic_schema_forbidden: "plan",
  query_type_mismatch: "plan",
  duplicate_output_column: "plan",
  invalid_aggregate: "plan",
  ast_depth_exceeded: "plan",
  variable_limit_exceeded: "plan",
  pattern_part_limit_exceeded: "plan",
  traversal_limit_exceeded: "plan",
  row_limit_exceeded: "plan",
  inaccessible_or_missing: "authorize",
  scope_not_granted: "authorize",
  foreign_draft: "authorize",
  query_cancelled: "execute",
  wall_time_exceeded: "execute",
  examined_expansions_exceeded: "execute",
  intermediate_rows_exceeded: "execute",
  invalid_engine_value: "normalize",
  heterogeneous_list: "normalize",
  integer_overflow: "normalize",
  duplicate_record_field: "normalize",
  result_row_too_large: "encode",
};

export function isGraphQueryErrorBody(body: GraphQueryErrorBody): body is {
  readonly error: {
    readonly code: GraphQueryCode;
    readonly phase: GraphQueryPhase;
    readonly path: string;
    readonly message?: string;
  };
} {
  const code = body.error?.code;
  if (typeof code !== "string" || !(code in QUERY_CODE_PHASES)) return false;
  const queryCode = code as GraphQueryCode;
  return body.error?.phase === QUERY_CODE_PHASES[queryCode]
    && typeof body.error.path === "string";
}
